content/IA-update_connections/connection-state-and-recovery.textile

```diff
@@ -0,0 +1,292 @@
+    NSLog(@"Ably is connected");
+  }];
+```
+
+```[swift]
+  let listener = realtime.connection.on(.connected) { change in
+    print("Ably is connected")
+  }
+```
+
+Alternatively a <span lang="default">listener</span><span lang="csharp">handler</span> may be registered so that it receives all state change events.
+
+```[jsall]
+  realtime.connection.on((stateChange) => {
+    console.log('New connection state is ' + stateChange.current);
+  });
+```
+
+```[java]
+  realtime.connection.on(new ConnectionStateListener() {
+    @Override
+    public void onConnectionStateChanged(ConnectionStateChange change) {
+      System.out.println("New state is " + change.current.name());
+    }
+  });
+```
+
+```[csharp]
+  realtime.Connection.On(args => {
+    Console.WriteLine("New state is " + args.Current)
+  });
+```
+
+```[ruby]
+  realtime.connection.on do |state_change|
+    puts "New connection state is #{state_change.current}"
+  end
+```
+
+```[objc]
+  ARTEventListener *listener = [realtime.connection on:^(ARTConnectionStateChange *change) {
+    NSLog(@"New connection state is %lu", (unsigned long)change.current);
+  }];
+```
+
+```[swift]
+  let listener = realtime.connection.on { change in
+    print("New connection state is \(change!.current)")
+  }
+```
+
+Previously registered <span lang="default">listeners</span><span lang="csharp">handlers</span> can be removed individually or all together.
+
+```[jsall]
+  /* remove a listener registered for a single event */
+  realtime.connection.off('connected', myListener);  
+
+  /* remove a listener registered for all events */
+  realtime.connection.off(myListener);  
+
+  /* remove all event listeners */
+  realtime.connection.off();
+```
+
+```[java]
+  /* remove a single listener */
+  realtime.connection.off(myListener);
+
+  /* remove all event listeners */
+  realtime.connection.off();
+```
+
+```[csharp]
+  /* remove a single handler */
+  realtime.Connection.Off(action);
+
+  /* remove all event handlers */
+  realtime.Connection.Off();
+```
+
+```[ruby]
+  # remove a listener registered for a single even
+  realtime.connection.off :connected, &block
+
+  # remove a listener registered for all events
+  realtime.connection.off &block
+
+  # remove all event listeners
+  realtime.connection.off
+```
+
+```[objc]
+  // remove a listener registered for a single event
+  [realtime.connection off:ARTRealtimeConnectionEventConnected listener:listener];
+
+  // remove a listener registered for all events
+  [realtime.connection off:listener];
+
+  // remove all event listeners
+  [realtime.connection off];
+```
+
+```[swift]
+  // remove a listener registered for a single event
+  realtime.connection.off(.connected, listener: listener)
+
+  // remove a listener registered for all events
+  realtime.connection.off(listener)
+
+  // remove all event listeners
+  realtime.connection.off()
+```
+
+h3(#handling-failures). Handling failures
```

Comment:
Should be h2 and "Handle connection failures".

Reply:
Fixed